File: src/main/java/org/moinex/ui/common/ProfitabilityMetricsPaneController.java
/*
 * Filename: ProfitabilityMetricsPaneController.java
 * Created on: February 21, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.common;

import java.math.BigDecimal;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import lombok.NoArgsConstructor;
import org.moinex.model.dto.ProfitabilityMetricsDTO;
import org.moinex.service.I18nService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public class ProfitabilityMetricsPaneController {

    @FXML private Label totalInvestedVariable;
    @FXML private Label totalInvestedFixed;
    @FXML private Label totalInvestedTotal;

    @FXML private Label currentValueVariable;
    @FXML private Label currentValueFixed;
    @FXML private Label currentValueTotal;

    @FXML private Label profitLossVariable;
    @FXML private Label profitLossFixed;
    @FXML private Label profitLossTotal;

    @FXML private Label returnPercentageVariable;
    @FXML private Label returnPercentageFixed;
    @FXML private Label returnPercentageTotal;

    @FXML private Label totalDividendsVariable;
    @FXML private Label totalDividendsFixed;
    @FXML private Label totalDividendsTotal;

    @FXML private Label dividendYieldVariable;
    @FXML private Label dividendYieldFixed;
    @FXML private Label dividendYieldTotal;

    private I18nService i18nService;

    @Autowired
    public ProfitabilityMetricsPaneController(I18nService i18nService) {
        this.i18nService = i18nService;
    }

    public void setMetrics(
            ProfitabilityMetricsDTO variableMetrics,
            ProfitabilityMetricsDTO fixedMetrics,
            ProfitabilityMetricsDTO totalMetrics) {
        setTotalInvested(
                variableMetrics.totalInvested(),
                fixedMetrics.totalInvested(),
                totalMetrics.totalInvested());
        setCurrentValue(
                variableMetrics.currentValue(),
                fixedMetrics.currentValue(),
                totalMetrics.currentValue());
        setProfitLoss(
                variableMetrics.profitLoss(), fixedMetrics.profitLoss(), totalMetrics.profitLoss());
        setReturnPercentage(
                variableMetrics.returnPercentage(),
                fixedMetrics.returnPercentage(),
                totalMetrics.returnPercentage());
        setTotalDividends(variableMetrics.totalDividends(), totalMetrics.totalDividends());
        setDividendYield(variableMetrics.dividendYield(), totalMetrics.dividendYield());
    }

    private void setTotalInvested(BigDecimal variable, BigDecimal fixed, BigDecimal total) {
        totalInvestedVariable.setText(UIUtils.formatCurrency(variable));
        totalInvestedFixed.setText(UIUtils.formatCurrency(fixed));
        totalInvestedTotal.setText(UIUtils.formatCurrency(total));
    }

    private void setCurrentValue(BigDecimal variable, BigDecimal fixed, BigDecimal total) {
        currentValueVariable.setText(UIUtils.formatCurrency(variable));
        currentValueFixed.setText(UIUtils.formatCurrency(fixed));
        currentValueTotal.setText(UIUtils.formatCurrency(total));
    }

    private void setProfitLoss(BigDecimal variable, BigDecimal fixed, BigDecimal total) {
        profitLossVariable.setText(formatValueWithSign(variable, false));
        addColorStyleClass(profitLossVariable, variable, true);

        profitLossFixed.setText(formatValueWithSign(fixed, false));
        addColorStyleClass(profitLossFixed, fixed, true);

        profitLossTotal.setText(formatValueWithSign(total, false));
        addColorStyleClass(profitLossTotal, total, true);
    }

    private void setReturnPercentage(BigDecimal variable, BigDecimal fixed, BigDecimal total) {
        returnPercentageVariable.setText(formatValueWithSign(variable, true));
        addColorStyleClass(returnPercentageVariable, variable, true);

        returnPercentageFixed.setText(formatValueWithSign(fixed, true));
        addColorStyleClass(returnPercentageFixed, fixed, true);

        returnPercentageTotal.setText(formatValueWithSign(total, true));
        addColorStyleClass(returnPercentageTotal, total, true);
    }

    private void setTotalDividends(BigDecimal variable, BigDecimal total) {
        totalDividendsVariable.setText(UIUtils.formatCurrency(variable));
        totalDividendsFixed.setText("-");
        totalDividendsFixed.getStyleClass().clear();
        totalDividendsFixed.getStyleClass().add(Constants.INFO_LABEL_NEUTRAL_STYLE);
        totalDividendsTotal.setText(UIUtils.formatCurrency(total));
    }

    private void setDividendYield(BigDecimal variable, BigDecimal total) {
        dividendYieldVariable.setText(UIUtils.formatPercentage(variable, i18nService));
        dividendYieldFixed.setText("-");
        dividendYieldFixed.getStyleClass().clear();
        dividendYieldFixed.getStyleClass().add(Constants.INFO_LABEL_NEUTRAL_STYLE);
        dividendYieldTotal.setText(UIUtils.formatPercentage(total, i18nService));
    }

    private String formatValueWithSign(BigDecimal value, boolean isPercentage) {
        String sign = "";
        if (value.compareTo(BigDecimal.ZERO) > 0) {
            sign = "+ ";
        } else if (value.compareTo(BigDecimal.ZERO) < 0) {
            sign = "- ";
        }

        String formattedValue =
                isPercentage
                        ? UIUtils.formatPercentage(value.abs(), i18nService)
                        : UIUtils.formatCurrency(value.abs());

        return sign + formattedValue;
    }

    private String getColorStyleClass(BigDecimal value, boolean dynamicColor) {
        if (!dynamicColor) {
            return Constants.INFO_LABEL_NEUTRAL_STYLE;
        }

        if (value.compareTo(BigDecimal.ZERO) < 0) {
            return Constants.INFO_LABEL_RED_STYLE;
        } else if (value.compareTo(BigDecimal.ZERO) > 0) {
            return Constants.INFO_LABEL_GREEN_STYLE;
        } else {
            return Constants.INFO_LABEL_NEUTRAL_STYLE;
        }
    }

    private void addColorStyleClass(Label label, BigDecimal value, boolean dynamicColor) {
        String colorClass = getColorStyleClass(value, dynamicColor);
        label.getStyleClass()
                .removeAll(
                        Constants.INFO_LABEL_RED_STYLE,
                        Constants.INFO_LABEL_GREEN_STYLE,
                        Constants.INFO_LABEL_NEUTRAL_STYLE);
        label.getStyleClass().add(colorClass);
    }
}

File: src/main/kotlin/org/moinex/ui/common/ProfitabilityMetricsPaneController.kt
/*
 * Filename: ProfitabilityMetricsPaneController.kt (original filename: ProfitabilityMetricsPaneController.java)
 * Created on: February 21, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 19/03/2026
 */

package org.moinex.ui.common

import javafx.fxml.FXML
import javafx.scene.control.Label
import org.moinex.common.constant.Styles
import org.moinex.common.util.UIUtils
import org.moinex.model.dto.ProfitabilityMetricsDTO
import org.springframework.stereotype.Controller
import java.math.BigDecimal

@Controller
class ProfitabilityMetricsPaneController {
    @FXML
    private lateinit var totalInvestedVariable: Label

    @FXML
    private lateinit var totalInvestedFixed: Label

    @FXML
    private lateinit var totalInvestedTotal: Label

    @FXML
    private lateinit var currentValueVariable: Label

    @FXML
    private lateinit var currentValueFixed: Label

    @FXML
    private lateinit var currentValueTotal: Label

    @FXML
    private lateinit var profitLossVariable: Label

    @FXML
    private lateinit var profitLossFixed: Label

    @FXML
    private lateinit var profitLossTotal: Label

    @FXML
    private lateinit var returnPercentageVariable: Label

    @FXML
    private lateinit var returnPercentageFixed: Label

    @FXML
    private lateinit var returnPercentageTotal: Label

    @FXML
    private lateinit var totalDividendsVariable: Label

    @FXML
    private lateinit var totalDividendsFixed: Label

    @FXML
    private lateinit var totalDividendsTotal: Label

    @FXML
    private lateinit var dividendYieldVariable: Label

    @FXML
    private lateinit var dividendYieldFixed: Label

    @FXML
    private lateinit var dividendYieldTotal: Label

    companion object {
        private const val DASH_PLACEHOLDER = "-"
        private const val POSITIVE_SIGN = "+ "
        private const val NEGATIVE_SIGN = "- "
    }

    fun setMetrics(
        variableMetrics: ProfitabilityMetricsDTO,
        fixedMetrics: ProfitabilityMetricsDTO,
        totalMetrics: ProfitabilityMetricsDTO,
    ) {
        setTotalInvested(
            variableMetrics.totalInvested,
            fixedMetrics.totalInvested,
            totalMetrics.totalInvested,
        )
        setCurrentValue(
            variableMetrics.currentValue,
            fixedMetrics.currentValue,
            totalMetrics.currentValue,
        )
        setProfitLoss(
            variableMetrics.profitLoss,
            fixedMetrics.profitLoss,
            totalMetrics.profitLoss,
        )
        setReturnPercentage(
            variableMetrics.returnPercentage,
            fixedMetrics.returnPercentage,
            totalMetrics.returnPercentage,
        )
        setTotalDividends(variableMetrics.totalDividends, totalMetrics.totalDividends)
        setDividendYield(variableMetrics.dividendYield, totalMetrics.dividendYield)
    }

    private fun setTotalInvested(
        variable: BigDecimal,
        fixed: BigDecimal,
        total: BigDecimal,
    ) {
        totalInvestedVariable.text = UIUtils.formatCurrency(variable)
        totalInvestedFixed.text = UIUtils.formatCurrency(fixed)
        totalInvestedTotal.text = UIUtils.formatCurrency(total)
    }

    private fun setCurrentValue(
        variable: BigDecimal,
        fixed: BigDecimal,
        total: BigDecimal,
    ) {
        currentValueVariable.text = UIUtils.formatCurrency(variable)
        currentValueFixed.text = UIUtils.formatCurrency(fixed)
        currentValueTotal.text = UIUtils.formatCurrency(total)
    }

    private fun setProfitLoss(
        variable: BigDecimal,
        fixed: BigDecimal,
        total: BigDecimal,
    ) {
        listOf(
            profitLossVariable to variable,
            profitLossFixed to fixed,
            profitLossTotal to total,
        ).forEach { (label, value) ->
            label.text = formatValueWithSign(value, isPercentage = false)
            addColorStyleClass(label, value)
        }
    }

    private fun setReturnPercentage(
        variable: BigDecimal,
        fixed: BigDecimal,
        total: BigDecimal,
    ) {
        listOf(
            returnPercentageVariable to variable,
            returnPercentageFixed to fixed,
            returnPercentageTotal to total,
        ).forEach { (label, value) ->
            label.text = formatValueWithSign(value, isPercentage = true)
            addColorStyleClass(label, value)
        }
    }

    private fun setTotalDividends(
        variable: BigDecimal,
        total: BigDecimal,
    ) {
        totalDividendsVariable.text = UIUtils.formatCurrency(variable)
        totalDividendsFixed.apply {
            text = DASH_PLACEHOLDER
            styleClass.apply {
                clear()
                add(Styles.INFO_LABEL_NEUTRAL_STYLE)
            }
        }
        totalDividendsTotal.text = UIUtils.formatCurrency(total)
    }

    private fun setDividendYield(
        variable: BigDecimal,
        total: BigDecimal,
    ) {
        dividendYieldVariable.text = UIUtils.formatPercentage(variable)
        dividendYieldFixed.apply {
            text = DASH_PLACEHOLDER
            styleClass.apply {
                clear()
                add(Styles.INFO_LABEL_NEUTRAL_STYLE)
            }
        }
        dividendYieldTotal.text = UIUtils.formatPercentage(total)
    }

    private fun formatValueWithSign(
        value: BigDecimal,
        isPercentage: Boolean,
    ): String {
        val sign =
            when {
                value > BigDecimal.ZERO -> POSITIVE_SIGN
                value < BigDecimal.ZERO -> NEGATIVE_SIGN
                else -> ""
            }

        val formattedValue =
            if (isPercentage) {
                UIUtils.formatPercentage(value.abs())
            } else {
                UIUtils.formatCurrency(value.abs())
            }

        return sign + formattedValue
    }

    private fun getColorStyleClass(
        value: BigDecimal,
        dynamicColor: Boolean,
    ): String =
        when {
            !dynamicColor -> Styles.INFO_LABEL_NEUTRAL_STYLE
            value < BigDecimal.ZERO -> Styles.INFO_LABEL_RED_STYLE
            value > BigDecimal.ZERO -> Styles.INFO_LABEL_GREEN_STYLE
            else -> Styles.INFO_LABEL_NEUTRAL_STYLE
        }

    private fun addColorStyleClass(
        label: Label,
        value: BigDecimal,
        dynamicColor: Boolean = true,
    ) {
        val colorClass = getColorStyleClass(value, dynamicColor)
        label.styleClass.apply {
            removeAll(
                Styles.INFO_LABEL_RED_STYLE,
                Styles.INFO_LABEL_GREEN_STYLE,
                Styles.INFO_LABEL_NEUTRAL_STYLE,
            )
            add(colorClass)
        }
    }
}
